# Capstone - Data Cleaning EPL and SPI Datasets

In [1]:
import numpy as np
import pandas as pd

## EPL Data Cleaning

In [2]:
pl_game_data = pd.read_csv('PL Database 1993-2018/final_premier_league.csv')

In [3]:
pl_game_data

Date           HomeTeam             AwayTeam  FTHG  FTAG FTR  \
0     1993-08-14         Arsenal FC        Coventry City     0     3   A   
1     1993-08-14        Aston Villa  Queens Park Rangers     4     1   H   
2     1993-08-14         Chelsea FC     Blackburn Rovers     1     2   A   
3     1993-08-14       Liverpool FC  Sheffield Wednesday     2     0   H   
4     1993-08-14    Manchester City         Leeds United     1     1   D   
...          ...                ...                  ...   ...   ...  ..   
9659  2018-05-13   Newcastle United           Chelsea FC     3     0   H   
9660  2018-05-13     Southampton FC      Manchester City     0     1   A   
9661  2018-05-13       Swansea City           Stoke City     1     2   A   
9662  2018-05-13  Tottenham Hotspur       Leicester City     5     4   H   
9663  2018-05-13    West Ham United           Everton FC     3     1   H   

       Season  Attendance  StadiumDistance  GoalDifference  ...    AF   HY  \
0     1993-94       30564               85              -3  ...   NaN  NaN   
1     1993-94       29015               98               3  ...   NaN  NaN   
2     1993-94       19416              183              -1  ...   NaN  NaN   
3     1993-94       38503               60               2  ...   NaN  NaN   
4     1993-94       26709               32               0  ...   NaN  NaN   
...       ...         ...              ...             ...  ...   ...  ...   
9659  2017-18       51992              248               3  ...  10.0  0.0   
9660  2017-18       30794              181              -1  ...  10.0  3.0   
9661  2017-18       20623              119              -1  ...   9.0  1.0   
9662  2017-18       67953               83               1  ...  13.0  1.0   
9663  2017-18       56885              182               2  ...  13.0  0.0   

       AY   HR   AR  FoulDifference  YellowDifference  RedDifference  \
0     NaN  NaN  NaN             NaN               NaN            NaN   
1     NaN  NaN  NaN             NaN               NaN            NaN   
2     NaN  NaN  NaN             NaN               NaN            NaN   
3     NaN  NaN  NaN             NaN               NaN            NaN   
4     NaN  NaN  NaN             NaN               NaN            NaN   
...   ...  ...  ...             ...               ...            ...   
9659  1.0  0.0  0.0            -1.0               1.0            0.0   
9660  1.0  0.0  0.0             2.0              -2.0            0.0   
9661  2.0  0.0  0.0            -3.0               1.0            0.0   
9662  2.0  0.0  0.0             4.0               1.0            0.0   
9663  1.0  0.0  0.0             3.0               1.0            0.0   

      RefBiasScore  EndingSeason  
0              NaN             -  
1              NaN             -  
2              NaN             -  
3              NaN             -  
4              NaN             -  
...            ...           ...  
9659      0.418338           Yes  
9660      0.436620           Yes  
9661      0.118557           Yes  
9662      0.453704           Yes  
9663      0.454545           Yes  

[9664 rows x 22 columns]

In [4]:
# Remove rows from up until 2009-2010 season
pl_game_data.loc[pl_game_data['Season'] == '2009-10']

# We see that the index for the first game of the 2009-2010 season is 6244, so we will remove all the rows above that.

Date                 HomeTeam         AwayTeam  FTHG  FTAG FTR  \
6244  2009-08-15              Aston Villa   Wigan Athletic     0     2   A   
6245  2009-08-15         Blackburn Rovers  Manchester City     0     2   A   
6246  2009-08-15         Bolton Wanderers   Sunderland AFC     0     1   A   
6247  2009-08-15               Chelsea FC        Hull City     2     1   H   
6248  2009-08-15               Everton FC       Arsenal FC     1     6   A   
...          ...                      ...              ...   ...   ...  ..   
6619  2010-05-09               Everton FC    Portsmouth FC     1     0   H   
6620  2010-05-09                Hull City     Liverpool FC     0     0   D   
6621  2010-05-09        Manchester United       Stoke City     4     0   H   
6622  2010-05-09          West Ham United  Manchester City     1     1   D   
6623  2010-05-09  Wolverhampton Wanderers   Sunderland AFC     2     1   H   

       Season  Attendance  StadiumDistance  GoalDifference  ...    AF   HY  \
6244  2009-10       38573               78              -2  ...  14.0  2.0   
6245  2009-10       25428               20              -2  ...   9.0  2.0   
6246  2009-10       21881              103              -1  ...  10.0  2.0   
6247  2009-10       41423              156               1  ...  15.0  1.0   
6248  2009-10       36725              177              -5  ...  13.0  0.0   
...       ...         ...              ...             ...  ...   ...  ...   
6619  2009-10       36725              199               1  ...  11.0  0.0   
6620  2009-10       24390              108               0  ...  11.0  1.0   
6621  2009-10       74864               33               4  ...   4.0  2.0   
6622  2009-10       33683              164               0  ...   7.0  2.0   
6623  2009-10       28366              163               1  ...  19.0  1.0   

       AY   HR   AR  FoulDifference  YellowDifference  RedDifference  \
6244  2.0  0.0  0.0            -1.0               0.0            0.0   
6245  1.0  0.0  0.0            -3.0              -1.0            0.0   
6246  1.0  0.0  0.0            -6.0              -1.0            0.0   
6247  2.0  0.0  0.0             2.0               1.0            0.0   
6248  0.0  0.0  0.0             2.0               0.0            0.0   
...   ...  ...  ...             ...               ...            ...   
6619  0.0  0.0  0.0            -1.0               0.0            0.0   
6620  0.0  0.0  0.0            -4.0              -1.0            0.0   
6621  0.0  0.0  0.0            -6.0              -2.0            0.0   
6622  0.0  0.0  0.0            -5.0              -2.0            0.0   
6623  2.0  0.0  2.0             8.0               1.0            2.0   

      RefBiasScore  EndingSeason  
6244      0.194539            No  
6245      0.258294            No  
6246      0.436620            No  
6247      0.425121            No  
6248      0.464912            No  
...            ...           ...  
6619      0.467456           Yes  
6620      0.436620           Yes  
6621      0.194539           Yes  
6622      0.491582           Yes  
6623      0.413223           Yes  

[380 rows x 22 columns]

In [5]:
# We see that the index for the first game of the 2009-2010 season is 6244, so we will remove all the rows above that.
pl_game_data = pl_game_data.drop(pl_game_data.index[0:6244])

In [6]:
pl_game_data = pl_game_data.reset_index(drop = True)

In [7]:
# Check NaN Values
pl_game_data.isna().sum()

Date                0
HomeTeam            0
AwayTeam            0
FTHG                0
FTAG                0
FTR                 0
Season              0
Attendance          0
StadiumDistance     0
GoalDifference      0
Referee             0
HF                  0
AF                  0
HY                  0
AY                  0
HR                  0
AR                  0
FoulDifference      0
YellowDifference    0
RedDifference       0
RefBiasScore        0
EndingSeason        0
dtype: int64

In [8]:
# Delete unwanted columns
pl_game_data.drop(columns = ['Attendance', 'StadiumDistance', 'FoulDifference', 'YellowDifference',\
                                      'RedDifference', 'EndingSeason'], inplace = True)

In [9]:
pl_game_data.head()

Date          HomeTeam         AwayTeam  FTHG  FTAG FTR   Season  \
0  2009-08-15       Aston Villa   Wigan Athletic     0     2   A  2009-10   
1  2009-08-15  Blackburn Rovers  Manchester City     0     2   A  2009-10   
2  2009-08-15  Bolton Wanderers   Sunderland AFC     0     1   A  2009-10   
3  2009-08-15        Chelsea FC        Hull City     2     1   H  2009-10   
4  2009-08-15        Everton FC       Arsenal FC     1     6   A  2009-10   

   GoalDifference        Referee    HF    AF   HY   AY   HR   AR  RefBiasScore  
0              -2  M Clattenburg  15.0  14.0  2.0  2.0  0.0  0.0      0.194539  
1              -2         M Dean  12.0   9.0  2.0  1.0  0.0  0.0      0.258294  
2              -1     A Marriner  16.0  10.0  2.0  1.0  0.0  0.0      0.436620  
3               1        A Wiley  13.0  15.0  1.0  2.0  0.0  0.0      0.425121  
4              -5       M Halsey  11.0  13.0  0.0  0.0  0.0  0.0      0.464912

In [10]:
pl_game_data.shape

(3420, 16)

In [11]:
pl_game_data0910 = pd.read_csv('PL Game Data By Season/season-0910.csv')

pl_game_data0910.head()

Div        Date     HomeTeam    AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  \
0  E0  2009-08-15  Aston Villa       Wigan     0     2   A     0     1   A   
1  E0  2009-08-15    Blackburn    Man City     0     2   A     0     1   A   
2  E0  2009-08-15       Bolton  Sunderland     0     1   A     0     1   A   
3  E0  2009-08-15      Chelsea        Hull     2     1   H     1     1   D   
4  E0  2009-08-15      Everton     Arsenal     1     6   A     0     3   A   

   ... BbMx>2.5  BbAv>2.5  BbMx<2.5  BbAv<2.5  BbAH  BbAHh  BbMxAHH  BbAvAHH  \
0  ...     2.04      1.92      1.91      1.81    21    0.0     1.28     1.22   
1  ...     2.15      2.04      1.78      1.71    24    0.0     2.58     2.38   
2  ...     2.25      2.12      1.70      1.66    23    0.0     1.68     1.61   
3  ...     1.67      1.58      2.44      2.26    22    0.0     1.03     1.02   
4  ...     2.20      2.06      1.77      1.70    24    0.0     2.27     2.20   

   BbMxAHA  BbAvAHA  
0     4.40     3.99  
1     1.60     1.54  
2     2.33     2.23  
3    17.05    12.96  
4     1.73     1.63  

[5 rows x 71 columns]

In [12]:
pl_game_data0910.shape

(380, 71)

In [13]:
pl_game_data1011 = pd.read_csv('PL Game Data By Season/season-1011.csv')

pl_game_data1011.shape

(380, 71)

In [14]:
pl_game_data1112 = pd.read_csv('PL Game Data By Season/season-1112.csv')

pl_game_data1112.shape

(380, 71)

In [15]:
pl_game_data1213 = pd.read_csv('PL Game Data By Season/season-1213.csv')

pl_game_data1213.shape

(380, 74)

In [16]:
pl_game_data1314 = pd.read_csv('PL Game Data By Season/season-1314.csv')

pl_game_data1314.shape

(380, 68)

In [17]:
pl_game_data1415 = pd.read_csv('PL Game Data By Season/season-1415.csv')

pl_game_data1415.shape # Why does this particular dataset have 1 extra column?  We need to investigate

(381, 68)

In [18]:
# Perhaps there are NaNs that we need to find.  Let's run .isna and .sum to aggregate any NaNs.
pl_game_data1415.isna().sum() # Clearly there is a row with missing values but which one?  We can use .loc to find it

Div         1
Date        1
HomeTeam    1
AwayTeam    1
FTHG        1
           ..
BbMxAHA     1
BbAvAHA     1
PSCH        1
PSCD        1
PSCA        1
Length: 68, dtype: int64

In [19]:
# We will use .loc to locate the row with missing data.  We can search in the 'Date' column for any NaN values.
pl_game_data1415.loc[pl_game_data1415['Date'].isna()] # The last row is blank so let's remove it from the dataframe.

Div Date HomeTeam AwayTeam  FTHG  FTAG  FTR  HTHG  HTAG  HTR  ...  \
380  NaN  NaN      NaN      NaN   NaN   NaN  NaN   NaN   NaN  NaN  ...   

    BbAv<2.5  BbAH  BbAHh  BbMxAHH  BbAvAHH  BbMxAHA  BbAvAHA  PSCH  PSCD  \
380      NaN   NaN    NaN      NaN      NaN      NaN      NaN   NaN   NaN   

     PSCA  
380   NaN  

[1 rows x 68 columns]

In [20]:
# Remove row 380 from the dataframe
pl_game_data1415 = pl_game_data1415.drop(pl_game_data1415.index[380])

In [21]:
# Check shape of pl_game_data1415 to see if we have 380 rows (i.e. the max total of 380 games per season)
pl_game_data1415.shape

(380, 68)

In [22]:
pl_game_data1516 = pd.read_csv('PL Game Data By Season/season-1516.csv')

pl_game_data1516.shape

(380, 65)

In [23]:
pl_game_data1617 = pd.read_csv('PL Game Data By Season/season-1617.csv')

pl_game_data1617.shape

(380, 65)

In [24]:
pl_game_data1718 = pd.read_csv('PL Game Data By Season/season-1718.csv')

pl_game_data1718.shape

(380, 65)

In [25]:
pl_game_data1819 = pd.read_csv('PL Game Data By Season/season-1819.csv')

pl_game_data1819.shape

(380, 62)

In [26]:
pl_game_data_concat = pd.concat([pl_game_data0910, pl_game_data1011, pl_game_data1112, pl_game_data1213,\
                                pl_game_data1314, pl_game_data1415, pl_game_data1516, pl_game_data1617,\
                                pl_game_data1718], axis = 0, sort = False)

# We will append pl_game_data1819 later because pl_game_data only has up to the 2017-2018 season

In [27]:
pl_game_data_concat.head()

Div        Date     HomeTeam    AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  \
0  E0  2009-08-15  Aston Villa       Wigan   0.0   2.0   A   0.0   1.0   A   
1  E0  2009-08-15    Blackburn    Man City   0.0   2.0   A   0.0   1.0   A   
2  E0  2009-08-15       Bolton  Sunderland   0.0   1.0   A   0.0   1.0   A   
3  E0  2009-08-15      Chelsea        Hull   2.0   1.0   H   1.0   1.0   D   
4  E0  2009-08-15      Everton     Arsenal   1.0   6.0   A   0.0   3.0   A   

   ... BbMxAHH  BbAvAHH  BbMxAHA  BbAvAHA  PSH  PSD  PSA  PSCH  PSCD  PSCA  
0  ...    1.28     1.22     4.40     3.99  NaN  NaN  NaN   NaN   NaN   NaN  
1  ...    2.58     2.38     1.60     1.54  NaN  NaN  NaN   NaN   NaN   NaN  
2  ...    1.68     1.61     2.33     2.23  NaN  NaN  NaN   NaN   NaN   NaN  
3  ...    1.03     1.02    17.05    12.96  NaN  NaN  NaN   NaN   NaN   NaN  
4  ...    2.27     2.20     1.73     1.63  NaN  NaN  NaN   NaN   NaN   NaN  

[5 rows x 77 columns]

In [28]:
# The above dataframe has valuable information but some of it is duplicated from dataset(pl_game_data). 
# Since this dataframe has most of the information that is found in the pl_data_data set, we will append the features
# defined in our wireframe to pl_game_data_concat.

In [29]:
# First we need to change the date format of pl_game_data to the same format as pl_game_data_concat

pl_game_data['Date'] = pd.to_datetime(pl_game_data['Date'])

pl_game_data_concat['Date'] = pd.to_datetime(pl_game_data_concat['Date'])

In [30]:
# Reset the index for pl_game_data_concat
pl_game_data_concat = pl_game_data_concat.reset_index(drop = True)

pl_game_data_concat

Div       Date     HomeTeam    AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  \
0     E0 2009-08-15  Aston Villa       Wigan   0.0   2.0   A   0.0   1.0   A   
1     E0 2009-08-15    Blackburn    Man City   0.0   2.0   A   0.0   1.0   A   
2     E0 2009-08-15       Bolton  Sunderland   0.0   1.0   A   0.0   1.0   A   
3     E0 2009-08-15      Chelsea        Hull   2.0   1.0   H   1.0   1.0   D   
4     E0 2009-08-15      Everton     Arsenal   1.0   6.0   A   0.0   3.0   A   
...   ..        ...          ...         ...   ...   ...  ..   ...   ...  ..   
3415  E0 2018-05-13    Newcastle     Chelsea   3.0   0.0   H   1.0   0.0   H   
3416  E0 2018-05-13  Southampton    Man City   0.0   1.0   A   0.0   0.0   D   
3417  E0 2018-05-13      Swansea       Stoke   1.0   2.0   A   1.0   2.0   A   
3418  E0 2018-05-13    Tottenham   Leicester   5.0   4.0   H   1.0   2.0   A   
3419  E0 2018-05-13     West Ham     Everton   3.0   1.0   H   1.0   0.0   H   

      ... BbMxAHH  BbAvAHH  BbMxAHA  BbAvAHA   PSH   PSD   PSA  PSCH  PSCD  \
0     ...    1.28     1.22     4.40     3.99   NaN   NaN   NaN   NaN   NaN   
1     ...    2.58     2.38     1.60     1.54   NaN   NaN   NaN   NaN   NaN   
2     ...    1.68     1.61     2.33     2.23   NaN   NaN   NaN   NaN   NaN   
3     ...    1.03     1.02    17.05    12.96   NaN   NaN   NaN   NaN   NaN   
4     ...    2.27     2.20     1.73     1.63   NaN   NaN   NaN   NaN   NaN   
...   ...     ...      ...      ...      ...   ...   ...   ...   ...   ...   
3415  ...    1.90     1.83     2.11     2.03  5.96  4.22  1.61  4.85  3.72   
3416  ...    2.01     1.95     1.97     1.91  8.06  5.11  1.42  6.32  4.78   
3417  ...    1.94     1.88     2.03     1.98  1.91  3.72  4.29  2.08  3.56   
3418  ...    1.96     1.86     2.05     2.00  1.33  5.95  9.56  1.38  5.50   
3419  ...    2.09     2.03     1.88     1.84  2.36  3.51  3.17  2.35  3.40   

      PSCA  
0      NaN  
1      NaN  
2      NaN  
3      NaN  
4      NaN  
...    ...  
3415  1.80  
3416  1.51  
3417  3.82  
3418  8.15  
3419  3.28  

[3420 rows x 77 columns]

In [31]:
# Next we can merge pl_game_data_concat with the features that we need from pl_game_data

# Print the columns from pl_game_data so we know which ones we want to add to pl_game_data_concat

pl_game_data.columns

Index(['Date', 'HomeTeam', 'AwayTeam', 'FTHG', 'FTAG', 'FTR', 'Season',
       'GoalDifference', 'Referee', 'HF', 'AF', 'HY', 'AY', 'HR', 'AR',
       'RefBiasScore'],
      dtype='object')

In [32]:
pl_game_data_concat = pd.merge(pl_game_data_concat, pl_game_data[['Date', 'HomeTeam', 'AwayTeam', 'Season', 'GoalDifference',\
                                                      'RefBiasScore']], left_index=True, right_index=True)

# We see that there is misalignment that is causing the right hand join to move up by 1 row.  This means that there
# is 1 row that is black.  We can use iloc to find that row, remove it and re-merge to see if everything aligns.

In [33]:
pl_game_data_concat

Div     Date_x   HomeTeam_x  AwayTeam_x  FTHG  FTAG FTR  HTHG  HTAG HTR  \
0     E0 2009-08-15  Aston Villa       Wigan   0.0   2.0   A   0.0   1.0   A   
1     E0 2009-08-15    Blackburn    Man City   0.0   2.0   A   0.0   1.0   A   
2     E0 2009-08-15       Bolton  Sunderland   0.0   1.0   A   0.0   1.0   A   
3     E0 2009-08-15      Chelsea        Hull   2.0   1.0   H   1.0   1.0   D   
4     E0 2009-08-15      Everton     Arsenal   1.0   6.0   A   0.0   3.0   A   
...   ..        ...          ...         ...   ...   ...  ..   ...   ...  ..   
3415  E0 2018-05-13    Newcastle     Chelsea   3.0   0.0   H   1.0   0.0   H   
3416  E0 2018-05-13  Southampton    Man City   0.0   1.0   A   0.0   0.0   D   
3417  E0 2018-05-13      Swansea       Stoke   1.0   2.0   A   1.0   2.0   A   
3418  E0 2018-05-13    Tottenham   Leicester   5.0   4.0   H   1.0   2.0   A   
3419  E0 2018-05-13     West Ham     Everton   3.0   1.0   H   1.0   0.0   H   

      ...   PSA  PSCH  PSCD  PSCA     Date_y         HomeTeam_y  \
0     ...   NaN   NaN   NaN   NaN 2009-08-15        Aston Villa   
1     ...   NaN   NaN   NaN   NaN 2009-08-15   Blackburn Rovers   
2     ...   NaN   NaN   NaN   NaN 2009-08-15   Bolton Wanderers   
3     ...   NaN   NaN   NaN   NaN 2009-08-15         Chelsea FC   
4     ...   NaN   NaN   NaN   NaN 2009-08-15         Everton FC   
...   ...   ...   ...   ...   ...        ...                ...   
3415  ...  1.61  4.85  3.72  1.80 2018-05-13   Newcastle United   
3416  ...  1.42  6.32  4.78  1.51 2018-05-13     Southampton FC   
3417  ...  4.29  2.08  3.56  3.82 2018-05-13       Swansea City   
3418  ...  9.56  1.38  5.50  8.15 2018-05-13  Tottenham Hotspur   
3419  ...  3.17  2.35  3.40  3.28 2018-05-13    West Ham United   

           AwayTeam_y   Season  GoalDifference  RefBiasScore  
0      Wigan Athletic  2009-10              -2      0.194539  
1     Manchester City  2009-10              -2      0.258294  
2      Sunderland AFC  2009-10              -1      0.436620  
3           Hull City  2009-10               1      0.425121  
4          Arsenal FC  2009-10              -5      0.464912  
...               ...      ...             ...           ...  
3415       Chelsea FC  2017-18               3      0.418338  
3416  Manchester City  2017-18              -1      0.436620  
3417       Stoke City  2017-18              -1      0.118557  
3418   Leicester City  2017-18               1      0.453704  
3419       Everton FC  2017-18               2      0.454545  

[3420 rows x 83 columns]

In [34]:
# Change the datetime format to pl_game_data1819 so that it can be ready to concatenate to pl_game_data_concat
pl_game_data1819['Date'] = pd.to_datetime(pl_game_data1819['Date'])

In [35]:
# Change Date_x, HomeTeam_x and AwayTeam_x so that pl_game_data1819 can be concatenated with NaN values
pl_game_data_concat = pl_game_data_concat.rename(columns = {'Date_x':'Date', 'HomeTeam_x':'HomeTeam',\
                                                            'AwayTeam_x':'AwayTeam'})

In [36]:
# Add pl_game_data1819 to the table
pl_game_data_concat = pd.concat([pl_game_data_concat, pl_game_data1819], axis = 0, sort = False, ignore_index = True)

In [37]:
# Now that we've appended the the correct features from pl_game_data to pl_game_data_concat, we can remove some of the
# features, namely the betting odds that we did not define in our original wireframe.

pl_game_data_concat.columns

Index(['Div', 'Date', 'HomeTeam', 'AwayTeam', 'FTHG', 'FTAG', 'FTR', 'HTHG',
       'HTAG', 'HTR', 'Referee', 'HS', 'AS', 'HST', 'AST', 'HF', 'AF', 'HC',
       'AC', 'HY', 'AY', 'HR', 'AR', 'B365H', 'B365D', 'B365A', 'BWH', 'BWD',
       'BWA', 'GBH', 'GBD', 'GBA', 'IWH', 'IWD', 'IWA', 'LBH', 'LBD', 'LBA',
       'SBH', 'SBD', 'SBA', 'WHH', 'WHD', 'WHA', 'SJH', 'SJD', 'SJA', 'VCH',
       'VCD', 'VCA', 'BSH', 'BSD', 'BSA', 'Bb1X2', 'BbMxH', 'BbAvH', 'BbMxD',
       'BbAvD', 'BbMxA', 'BbAvA', 'BbOU', 'BbMx>2.5', 'BbAv>2.5', 'BbMx<2.5',
       'BbAv<2.5', 'BbAH', 'BbAHh', 'BbMxAHH', 'BbAvAHH', 'BbMxAHA', 'BbAvAHA',
       'PSH', 'PSD', 'PSA', 'PSCH', 'PSCD', 'PSCA', 'Date_y', 'HomeTeam_y',
       'AwayTeam_y', 'Season', 'GoalDifference', 'RefBiasScore'],
      dtype='object')

In [38]:
# Drop Unnecessary columns
pl_game_data_concat = pl_game_data_concat.drop(columns = ['BWH','BWD', 'BWA', 'GBH', 'GBD', 'GBA', 'IWH', 'IWD',\
                                                          'IWA', 'SJD', 'SJA', 'VCH', 'VCD', 'VCA', 'BSH', 'BSD',\
                                                          'BSA', 'Bb1X2', 'BbMxH', 'BbAvH', 'BbMxD', 'BbAvD',\
                                                          'BbMxA', 'BbAvA', 'BbOU', 'BbMx>2.5', 'BbAv>2.5',\
                                                          'BbMx<2.5', 'BbAv<2.5', 'BbAH', 'BbAHh', 'BbMxAHH',\
                                                          'BbAvAHH', 'BbMxAHA', 'BbAvAHA', 'PSH', 'PSD', 'PSA',\
                                                          'PSCH', 'PSCD', 'PSCA'])

In [39]:
pl_game_data_concat

Div       Date     HomeTeam      AwayTeam  FTHG  FTAG FTR  HTHG  HTAG  \
0     E0 2009-08-15  Aston Villa         Wigan   0.0   2.0   A   0.0   1.0   
1     E0 2009-08-15    Blackburn      Man City   0.0   2.0   A   0.0   1.0   
2     E0 2009-08-15       Bolton    Sunderland   0.0   1.0   A   0.0   1.0   
3     E0 2009-08-15      Chelsea          Hull   2.0   1.0   H   1.0   1.0   
4     E0 2009-08-15      Everton       Arsenal   1.0   6.0   A   0.0   3.0   
...   ..        ...          ...           ...   ...   ...  ..   ...   ...   
3795  E0 2019-12-05    Liverpool        Wolves   2.0   0.0   H   1.0   0.0   
3796  E0 2019-12-05   Man United       Cardiff   0.0   2.0   A   0.0   1.0   
3797  E0 2019-12-05  Southampton  Huddersfield   1.0   1.0   D   1.0   0.0   
3798  E0 2019-12-05    Tottenham       Everton   2.0   2.0   D   1.0   0.0   
3799  E0 2019-12-05      Watford      West Ham   1.0   4.0   A   0.0   2.0   

     HTR  ...   WHH   WHD    WHA   SJH     Date_y        HomeTeam_y  \
0      A  ...  1.70  3.40   5.50  1.67 2009-08-15       Aston Villa   
1      A  ...  3.50  3.20   2.15  3.30 2009-08-15  Blackburn Rovers   
2      A  ...  2.30  3.20   3.20  2.30 2009-08-15  Bolton Wanderers   
3      D  ...  1.17  6.50  21.00  1.18 2009-08-15        Chelsea FC   
4      A  ...  3.20  3.20   2.30  3.25 2009-08-15        Everton FC   
...   ..  ...   ...   ...    ...   ...        ...               ...   
3795   H  ...  1.29  5.50  11.00   NaN        NaT               NaN   
3796   A  ...  1.25  6.50  10.00   NaN        NaT               NaN   
3797   H  ...  1.40  4.75   8.00   NaN        NaT               NaN   
3798   H  ...  2.10  3.50   3.50   NaN        NaT               NaN   
3799   A  ...  2.20  3.70   3.10   NaN        NaT               NaN   

           AwayTeam_y   Season  GoalDifference  RefBiasScore  
0      Wigan Athletic  2009-10            -2.0      0.194539  
1     Manchester City  2009-10            -2.0      0.258294  
2      Sunderland AFC  2009-10            -1.0      0.436620  
3           Hull City  2009-10             1.0      0.425121  
4          Arsenal FC  2009-10            -5.0      0.464912  
...               ...      ...             ...           ...  
3795              NaN      NaN             NaN           NaN  
3796              NaN      NaN             NaN           NaN  
3797              NaN      NaN             NaN           NaN  
3798              NaN      NaN             NaN           NaN  
3799              NaN      NaN             NaN           NaN  

[3800 rows x 42 columns]

In [40]:
# Check for NaNs
pl_game_data_concat.isna().sum()

Div                  0
Date                 0
HomeTeam             0
AwayTeam             0
FTHG                 0
FTAG                 0
FTR                  0
HTHG                 0
HTAG                 0
HTR                  0
Referee              0
HS                   0
AS                   0
HST                  0
AST                  0
HF                   0
AF                   0
HC                   0
AC                   0
HY                   0
AY                   0
HR                   0
AR                   0
B365H                0
B365D                0
B365A                0
LBH                381
LBD                381
LBA                381
SBH               2660
SBD               2660
SBA               2660
WHH                  0
WHD                  0
WHA                  0
SJH               1860
Date_y             380
HomeTeam_y         380
AwayTeam_y         380
Season             380
GoalDifference     380
RefBiasScore       380
dtype: int64

In [41]:
# Drop unnecessary columns
pl_game_data_concat = pl_game_data_concat.drop(columns = ['HTHG', 'HTAG', 'Referee', 'LBH', 'LBD', 'LBA', 'SBH',\
                                                          'LBD', 'LBA', 'SBH', 'SBD', 'SBA', 'SBD', 'SBA', 'WHH',\
                                                          'WHD', 'WHA', 'SJH', 'Date_y', 'HomeTeam_y', 'AwayTeam_y',\
                                                          'RefBiasScore', 'Div'])

In [42]:
# Sort the dataframe for legibility purposes
pl_game_data_concat = pl_game_data_concat[['Date', 'Season', 'HomeTeam', 'AwayTeam', 'FTHG', 'FTAG',\
                                           'GoalDifference','HS', 'AS', 'HST', 'AST', 'HF', 'AF', 'HC', 'AC', 'HY',\
                                           'AY', 'HR', 'AR', 'B365H', 'B365D', 'B365A', 'FTR']]

In [43]:
pl_game_data_concat.head()

Date   Season     HomeTeam    AwayTeam  FTHG  FTAG  GoalDifference  \
0 2009-08-15  2009-10  Aston Villa       Wigan   0.0   2.0            -2.0   
1 2009-08-15  2009-10    Blackburn    Man City   0.0   2.0            -2.0   
2 2009-08-15  2009-10       Bolton  Sunderland   0.0   1.0            -1.0   
3 2009-08-15  2009-10      Chelsea        Hull   2.0   1.0             1.0   
4 2009-08-15  2009-10      Everton     Arsenal   1.0   6.0            -5.0   

     HS    AS   HST  ...    HC   AC   HY   AY   HR   AR  B365H  B365D  B365A  \
0  11.0  14.0   5.0  ...   4.0  6.0  2.0  2.0  0.0  0.0   1.67   3.60   5.50   
1  17.0   8.0   9.0  ...   5.0  4.0  2.0  1.0  0.0  0.0   3.60   3.25   2.10   
2  11.0  20.0   3.0  ...   4.0  7.0  2.0  1.0  0.0  0.0   2.25   3.25   3.25   
3  26.0   7.0  12.0  ...  12.0  4.0  1.0  2.0  0.0  0.0   1.17   6.50  21.00   
4   8.0  15.0   5.0  ...   4.0  9.0  0.0  0.0  0.0  0.0   3.20   3.25   2.30   

   FTR  
0    A  
1    A  
2    A  
3    H  
4    A  

[5 rows x 23 columns]

In [44]:
pl_game_data_concat.isna().sum()

Date                0
Season            380
HomeTeam            0
AwayTeam            0
FTHG                0
FTAG                0
GoalDifference    380
HS                  0
AS                  0
HST                 0
AST                 0
HF                  0
AF                  0
HC                  0
AC                  0
HY                  0
AY                  0
HR                  0
AR                  0
B365H               0
B365D               0
B365A               0
FTR                 0
dtype: int64

In [45]:
# I know that the 2018-2019 season wasn't filled in, so we will use this to fill the NaNs.
pl_game_data_concat['Season'].fillna('2018-2019', inplace = True)

In [46]:
# We can calculate the missing values for goal difference by subtracting the Home Goals with the Away Goals
pl_game_data_concat['GoalDifference'].fillna((pl_game_data_concat['FTHG']-pl_game_data_concat['FTAG']), inplace = True)

In [47]:
pl_game_data_concat.isna().sum()

Date              0
Season            0
HomeTeam          0
AwayTeam          0
FTHG              0
FTAG              0
GoalDifference    0
HS                0
AS                0
HST               0
AST               0
HF                0
AF                0
HC                0
AC                0
HY                0
AY                0
HR                0
AR                0
B365H             0
B365D             0
B365A             0
FTR               0
dtype: int64

In [48]:
pl_game_data_concat.dtypes

Date              datetime64[ns]
Season                    object
HomeTeam                  object
AwayTeam                  object
FTHG                     float64
FTAG                     float64
GoalDifference           float64
HS                       float64
AS                       float64
HST                      float64
AST                      float64
HF                       float64
AF                       float64
HC                       float64
AC                       float64
HY                       float64
AY                       float64
HR                       float64
AR                       float64
B365H                    float64
B365D                    float64
B365A                    float64
FTR                       object
dtype: object

In [49]:
# We will use this to conduct the poisson distribution for the EPL data only.
pl_game_data_concat.to_csv(r'pl_data_clean.csv')

----

# SPI Cleaning

In [50]:
# Import the CSV
pl_spi = pd.read_csv('spi_matches.csv')

pl_spi

date  league_id                   league           team1  \
0      2016-08-12       1843           French Ligue 1          Bastia   
1      2016-08-12       1843           French Ligue 1       AS Monaco   
2      2016-08-13       2411  Barclays Premier League       Hull City   
3      2016-08-13       2411  Barclays Premier League  Crystal Palace   
4      2016-08-13       2411  Barclays Premier League         Everton   
...           ...        ...                      ...             ...   
34129  2020-11-29       1859    Norwegian Tippeligaen       Viking FK   
34130  2020-11-29       1859    Norwegian Tippeligaen        SK Brann   
34131  2020-11-29       1859    Norwegian Tippeligaen       Haugesund   
34132  2020-11-29       1859    Norwegian Tippeligaen       Mjondalen   
34133  2020-11-29       1859    Norwegian Tippeligaen           Molde   

                      team2   spi1   spi2   prob1   prob2  probtie  ...  \
0       Paris Saint-Germain  51.16  85.68  0.0463  0.8380   0.1157  ...   
1                  Guingamp  68.85  56.48  0.5714  0.1669   0.2617  ...   
2            Leicester City  53.57  66.81  0.3459  0.3621   0.2921  ...   
3      West Bromwich Albion  55.19  58.66  0.4214  0.2939   0.2847  ...   
4         Tottenham Hotspur  68.02  73.25  0.3910  0.3401   0.2689  ...   
...                     ...    ...    ...     ...     ...      ...  ...   
34129          Stromsgodset  32.76  25.85  0.5548  0.2232   0.2220  ...   
34130              Aalesund  25.72  20.51  0.5343  0.1995   0.2662  ...   
34131            Sandefjord  30.35  17.15  0.6259  0.1435   0.2306  ...   
34132             Valerenga  20.46  25.91  0.3968  0.3384   0.2648  ...   
34133            Bodo/Glimt  53.66  35.87  0.6580  0.1566   0.1853  ...   

       importance1  importance2  score1  score2   xg1   xg2  nsxg1  nsxg2  \
0             32.4         67.7     0.0     1.0  0.97  0.63   0.43   0.45   
1             53.7         22.9     2.0     2.0  2.45  0.77   1.75   0.42   
2             38.1         22.2     2.0     1.0  0.85  2.77   0.17   1.25   
3             43.6         34.6     0.0     1.0  1.11  0.68   0.84   1.60   
4             31.9         48.0     1.0     1.0  0.73  1.11   0.88   1.81   
...            ...          ...     ...     ...   ...   ...    ...    ...   
34129          NaN          NaN     NaN     NaN   NaN   NaN    NaN    NaN   
34130          NaN          NaN     NaN     NaN   NaN   NaN    NaN    NaN   
34131          NaN          NaN     NaN     NaN   NaN   NaN    NaN    NaN   
34132          NaN          NaN     NaN     NaN   NaN   NaN    NaN    NaN   
34133          NaN          NaN     NaN     NaN   NaN   NaN    NaN    NaN   

       adj_score1  adj_score2  
0            0.00        1.05  
1            2.10        2.10  
2            2.10        1.05  
3            0.00        1.05  
4            1.05        1.05  
...           ...         ...  
34129         NaN         NaN  
34130         NaN         NaN  
34131         NaN         NaN  
34132         NaN         NaN  
34133         NaN         NaN  

[34134 rows x 22 columns]

In [51]:
# We see that there are over 34,000 rows and that there are multiple leagues.  We only want to pull the Barclays
# Premier League data so we will only pull these rows.

pl_spi = pl_spi.loc[pl_spi['league'] == 'Barclays Premier League']

In [52]:
pl_spi

date  league_id                   league            team1  \
2      2016-08-13       2411  Barclays Premier League        Hull City   
3      2016-08-13       2411  Barclays Premier League   Crystal Palace   
4      2016-08-13       2411  Barclays Premier League          Everton   
5      2016-08-13       2411  Barclays Premier League    Middlesbrough   
6      2016-08-13       2411  Barclays Premier League          Burnley   
...           ...        ...                      ...              ...   
32894  2020-05-17       2411  Barclays Premier League  West Ham United   
32895  2020-05-17       2411  Barclays Premier League        Newcastle   
32896  2020-05-17       2411  Barclays Premier League  Manchester City   
32897  2020-05-17       2411  Barclays Premier League          Arsenal   
32899  2020-05-17       2411  Barclays Premier League          Burnley   

                          team2   spi1   spi2   prob1   prob2  probtie  ...  \
2                Leicester City  53.57  66.81  0.3459  0.3621   0.2921  ...   
3          West Bromwich Albion  55.19  58.66  0.4214  0.2939   0.2847  ...   
4             Tottenham Hotspur  68.02  73.25  0.3910  0.3401   0.2689  ...   
5                    Stoke City  56.32  60.35  0.4380  0.2692   0.2927  ...   
6                  Swansea City  58.98  59.74  0.4482  0.2663   0.2854  ...   
...                         ...    ...    ...     ...     ...      ...  ...   
32894               Aston Villa  66.68  60.54  0.5202  0.2540   0.2258  ...   
32895                 Liverpool  60.18  92.67  0.0984  0.7342   0.1674  ...   
32896              Norwich City  95.12  61.04  0.9197  0.0160   0.0643  ...   
32897                   Watford  77.04  72.94  0.5111  0.2446   0.2443  ...   
32899  Brighton and Hove Albion  70.05  69.59  0.4455  0.2848   0.2697  ...   

       importance1  importance2  score1  score2   xg1   xg2  nsxg1  nsxg2  \
2             38.1         22.2     2.0     1.0  0.85  2.77   0.17   1.25   
3             43.6         34.6     0.0     1.0  1.11  0.68   0.84   1.60   
4             31.9         48.0     1.0     1.0  0.73  1.11   0.88   1.81   
5             33.9         32.5     1.0     1.0  1.40  0.55   1.13   1.06   
6             36.5         29.1     0.0     1.0  1.24  1.84   1.71   1.56   
...            ...          ...     ...     ...   ...   ...    ...    ...   
32894          NaN          NaN     NaN     NaN   NaN   NaN    NaN    NaN   
32895          NaN          NaN     NaN     NaN   NaN   NaN    NaN    NaN   
32896          NaN          NaN     NaN     NaN   NaN   NaN    NaN    NaN   
32897          NaN          NaN     NaN     NaN   NaN   NaN    NaN    NaN   
32899          NaN          NaN     NaN     NaN   NaN   NaN    NaN    NaN   

       adj_score1  adj_score2  
2            2.10        1.05  
3            0.00        1.05  
4            1.05        1.05  
5            1.05        1.05  
6            0.00        1.05  
...           ...         ...  
32894         NaN         NaN  
32895         NaN         NaN  
32896         NaN         NaN  
32897         NaN         NaN  
32899         NaN         NaN  

[1520 rows x 22 columns]

In [53]:
pl_spi.columns

Index(['date', 'league_id', 'league', 'team1', 'team2', 'spi1', 'spi2',
       'prob1', 'prob2', 'probtie', 'proj_score1', 'proj_score2',
       'importance1', 'importance2', 'score1', 'score2', 'xg1', 'xg2', 'nsxg1',
       'nsxg2', 'adj_score1', 'adj_score2'],
      dtype='object')

In [54]:
# we need to rename the columns in pl_spi and change the datatype

pl_spi = pl_spi.rename(columns = {'date':'Date', 'team1':'HomeTeam', 'team2':'AwayTeam', 'spi1':'HomeSPI',\
                                  'spi2':'AwaySPI', 'prob1':'HomeWinProb', 'prob2':'AwayWinProb',\
                                  'probtie':'TieProb', 'proj_score1':'HomeProjScore', 'proj_score2':'AwayProjScore',\
                                  'importance1':'HomeImportance', 'importance2':'AwayImportance',\
                                  'xg1':'HomeShotXPG', 'xg2':'AwayShotXPG', 'nsxg1':'HomeNonShotXPG',\
                                  'nsxg2':'AwayNonShotXPG', 'adj_score1':'HomeAdjScore', 'adj_score2':'AwayAdjScore'})

pl_spi['Date'] = pd.to_datetime(pl_spi['Date'])

In [55]:
pl_spi.head()

Date  league_id                   league        HomeTeam  \
2 2016-08-13       2411  Barclays Premier League       Hull City   
3 2016-08-13       2411  Barclays Premier League  Crystal Palace   
4 2016-08-13       2411  Barclays Premier League         Everton   
5 2016-08-13       2411  Barclays Premier League   Middlesbrough   
6 2016-08-13       2411  Barclays Premier League         Burnley   

               AwayTeam  HomeSPI  AwaySPI  HomeWinProb  AwayWinProb  TieProb  \
2        Leicester City    53.57    66.81       0.3459       0.3621   0.2921   
3  West Bromwich Albion    55.19    58.66       0.4214       0.2939   0.2847   
4     Tottenham Hotspur    68.02    73.25       0.3910       0.3401   0.2689   
5            Stoke City    56.32    60.35       0.4380       0.2692   0.2927   
6          Swansea City    58.98    59.74       0.4482       0.2663   0.2854   

   ...  HomeImportance  AwayImportance  score1  score2  HomeShotXPG  \
2  ...            38.1            22.2     2.0     1.0         0.85   
3  ...            43.6            34.6     0.0     1.0         1.11   
4  ...            31.9            48.0     1.0     1.0         0.73   
5  ...            33.9            32.5     1.0     1.0         1.40   
6  ...            36.5            29.1     0.0     1.0         1.24   

   AwayShotXPG  HomeNonShotXPG  AwayNonShotXPG  HomeAdjScore  AwayAdjScore  
2         2.77            0.17            1.25          2.10          1.05  
3         0.68            0.84            1.60          0.00          1.05  
4         1.11            0.88            1.81          1.05          1.05  
5         0.55            1.13            1.06          1.05          1.05  
6         1.84            1.71            1.56          0.00          1.05  

[5 rows x 22 columns]

In [56]:
# Drop unwanted columns
pl_spi = pl_spi.drop(columns = ['league_id', 'league'])

In [57]:
pl_spi

Date         HomeTeam                  AwayTeam  HomeSPI  AwaySPI  \
2     2016-08-13        Hull City            Leicester City    53.57    66.81   
3     2016-08-13   Crystal Palace      West Bromwich Albion    55.19    58.66   
4     2016-08-13          Everton         Tottenham Hotspur    68.02    73.25   
5     2016-08-13    Middlesbrough                Stoke City    56.32    60.35   
6     2016-08-13          Burnley              Swansea City    58.98    59.74   
...          ...              ...                       ...      ...      ...   
32894 2020-05-17  West Ham United               Aston Villa    66.68    60.54   
32895 2020-05-17        Newcastle                 Liverpool    60.18    92.67   
32896 2020-05-17  Manchester City              Norwich City    95.12    61.04   
32897 2020-05-17          Arsenal                   Watford    77.04    72.94   
32899 2020-05-17          Burnley  Brighton and Hove Albion    70.05    69.59   

       HomeWinProb  AwayWinProb  TieProb  HomeProjScore  AwayProjScore  \
2           0.3459       0.3621   0.2921           1.16           1.24   
3           0.4214       0.2939   0.2847           1.35           1.14   
4           0.3910       0.3401   0.2689           1.47           1.38   
5           0.4380       0.2692   0.2927           1.30           1.01   
6           0.4482       0.2663   0.2854           1.37           1.05   
...            ...          ...      ...            ...            ...   
32894       0.5202       0.2540   0.2258           1.95           1.31   
32895       0.0984       0.7342   0.1674           0.76           2.40   
32896       0.9197       0.0160   0.0643           3.33           0.33   
32897       0.5111       0.2446   0.2443           1.72           1.12   
32899       0.4455       0.2848   0.2697           1.44           1.10   

       HomeImportance  AwayImportance  score1  score2  HomeShotXPG  \
2                38.1            22.2     2.0     1.0         0.85   
3                43.6            34.6     0.0     1.0         1.11   
4                31.9            48.0     1.0     1.0         0.73   
5                33.9            32.5     1.0     1.0         1.40   
6                36.5            29.1     0.0     1.0         1.24   
...               ...             ...     ...     ...          ...   
32894             NaN             NaN     NaN     NaN          NaN   
32895             NaN             NaN     NaN     NaN          NaN   
32896             NaN             NaN     NaN     NaN          NaN   
32897             NaN             NaN     NaN     NaN          NaN   
32899             NaN             NaN     NaN     NaN          NaN   

       AwayShotXPG  HomeNonShotXPG  AwayNonShotXPG  HomeAdjScore  AwayAdjScore  
2             2.77            0.17            1.25          2.10          1.05  
3             0.68            0.84            1.60          0.00          1.05  
4             1.11            0.88            1.81          1.05          1.05  
5             0.55            1.13            1.06          1.05          1.05  
6             1.84            1.71            1.56          0.00          1.05  
...            ...             ...             ...           ...           ...  
32894          NaN             NaN             NaN           NaN           NaN  
32895          NaN             NaN             NaN           NaN           NaN  
32896          NaN             NaN             NaN           NaN           NaN  
32897          NaN             NaN             NaN           NaN           NaN  
32899          NaN             NaN             NaN           NaN           NaN  

[1520 rows x 20 columns]

In [58]:
# Reset the index
pl_spi = pl_spi.reset_index(drop = True)

In [59]:
pl_spi

Date         HomeTeam                  AwayTeam  HomeSPI  AwaySPI  \
0    2016-08-13        Hull City            Leicester City    53.57    66.81   
1    2016-08-13   Crystal Palace      West Bromwich Albion    55.19    58.66   
2    2016-08-13          Everton         Tottenham Hotspur    68.02    73.25   
3    2016-08-13    Middlesbrough                Stoke City    56.32    60.35   
4    2016-08-13          Burnley              Swansea City    58.98    59.74   
...         ...              ...                       ...      ...      ...   
1515 2020-05-17  West Ham United               Aston Villa    66.68    60.54   
1516 2020-05-17        Newcastle                 Liverpool    60.18    92.67   
1517 2020-05-17  Manchester City              Norwich City    95.12    61.04   
1518 2020-05-17          Arsenal                   Watford    77.04    72.94   
1519 2020-05-17          Burnley  Brighton and Hove Albion    70.05    69.59   

      HomeWinProb  AwayWinProb  TieProb  HomeProjScore  AwayProjScore  \
0          0.3459       0.3621   0.2921           1.16           1.24   
1          0.4214       0.2939   0.2847           1.35           1.14   
2          0.3910       0.3401   0.2689           1.47           1.38   
3          0.4380       0.2692   0.2927           1.30           1.01   
4          0.4482       0.2663   0.2854           1.37           1.05   
...           ...          ...      ...            ...            ...   
1515       0.5202       0.2540   0.2258           1.95           1.31   
1516       0.0984       0.7342   0.1674           0.76           2.40   
1517       0.9197       0.0160   0.0643           3.33           0.33   
1518       0.5111       0.2446   0.2443           1.72           1.12   
1519       0.4455       0.2848   0.2697           1.44           1.10   

      HomeImportance  AwayImportance  score1  score2  HomeShotXPG  \
0               38.1            22.2     2.0     1.0         0.85   
1               43.6            34.6     0.0     1.0         1.11   
2               31.9            48.0     1.0     1.0         0.73   
3               33.9            32.5     1.0     1.0         1.40   
4               36.5            29.1     0.0     1.0         1.24   
...              ...             ...     ...     ...          ...   
1515             NaN             NaN     NaN     NaN          NaN   
1516             NaN             NaN     NaN     NaN          NaN   
1517             NaN             NaN     NaN     NaN          NaN   
1518             NaN             NaN     NaN     NaN          NaN   
1519             NaN             NaN     NaN     NaN          NaN   

      AwayShotXPG  HomeNonShotXPG  AwayNonShotXPG  HomeAdjScore  AwayAdjScore  
0            2.77            0.17            1.25          2.10          1.05  
1            0.68            0.84            1.60          0.00          1.05  
2            1.11            0.88            1.81          1.05          1.05  
3            0.55            1.13            1.06          1.05          1.05  
4            1.84            1.71            1.56          0.00          1.05  
...           ...             ...             ...           ...           ...  
1515          NaN             NaN             NaN           NaN           NaN  
1516          NaN             NaN             NaN           NaN           NaN  
1517          NaN             NaN             NaN           NaN           NaN  
1518          NaN             NaN             NaN           NaN           NaN  
1519          NaN             NaN             NaN           NaN           NaN  

[1520 rows x 20 columns]

In [60]:
pl_spi.isna().sum()

Date                0
HomeTeam            0
AwayTeam            0
HomeSPI             0
AwaySPI             0
HomeWinProb         0
AwayWinProb         0
TieProb             0
HomeProjScore       0
AwayProjScore       0
HomeImportance    126
AwayImportance    126
score1            102
score2            102
HomeShotXPG       102
AwayShotXPG       102
HomeNonShotXPG    102
AwayNonShotXPG    102
HomeAdjScore      102
AwayAdjScore      102
dtype: int64

In [61]:
# We can see that we have 102 NaN values which are games that have not been played yet.  Let's remove those games.
# We have 1520-102 = 1418
pl_spi = pl_spi.drop(pl_spi.index[1418:1520])

In [62]:
pl_spi

Date         HomeTeam              AwayTeam  HomeSPI  AwaySPI  \
0    2016-08-13        Hull City        Leicester City    53.57    66.81   
1    2016-08-13   Crystal Palace  West Bromwich Albion    55.19    58.66   
2    2016-08-13          Everton     Tottenham Hotspur    68.02    73.25   
3    2016-08-13    Middlesbrough            Stoke City    56.32    60.35   
4    2016-08-13          Burnley          Swansea City    58.98    59.74   
...         ...              ...                   ...      ...      ...   
1413 2020-02-29      Aston Villa      Sheffield United    60.54    73.71   
1414 2020-02-29  West Ham United           Southampton    65.53    71.96   
1415 2020-02-29  AFC Bournemouth               Chelsea    62.54    84.59   
1416 2020-02-29          Watford             Liverpool    70.07    93.98   
1417 2020-03-01  Manchester City               Arsenal    95.12    77.04   

      HomeWinProb  AwayWinProb  TieProb  HomeProjScore  AwayProjScore  \
0          0.3459       0.3621   0.2921           1.16           1.24   
1          0.4214       0.2939   0.2847           1.35           1.14   
2          0.3910       0.3401   0.2689           1.47           1.38   
3          0.4380       0.2692   0.2927           1.30           1.01   
4          0.4482       0.2663   0.2854           1.37           1.05   
...           ...          ...      ...            ...            ...   
1413       0.3630       0.3699   0.2671           1.32           1.34   
1414       0.4304       0.3190   0.2507           1.59           1.34   
1415       0.2047       0.5733   0.2220           1.10           1.98   
1416       0.1706       0.6234   0.2061           1.02           2.13   
1417       0.7885       0.0761   0.1353           2.79           0.77   

      HomeImportance  AwayImportance  score1  score2  HomeShotXPG  \
0               38.1            22.2     2.0     1.0         0.85   
1               43.6            34.6     0.0     1.0         1.11   
2               31.9            48.0     1.0     1.0         0.73   
3               33.9            32.5     1.0     1.0         1.40   
4               36.5            29.1     0.0     1.0         1.24   
...              ...             ...     ...     ...          ...   
1413            62.2            39.4     NaN     NaN          NaN   
1414            50.3             4.9     3.0     1.0         2.63   
1415            66.6            78.8     2.0     2.0         1.55   
1416            57.7             0.0     3.0     0.0         2.31   
1417             0.1            37.0     NaN     NaN          NaN   

      AwayShotXPG  HomeNonShotXPG  AwayNonShotXPG  HomeAdjScore  AwayAdjScore  
0            2.77            0.17            1.25          2.10          1.05  
1            0.68            0.84            1.60          0.00          1.05  
2            1.11            0.88            1.81          1.05          1.05  
3            0.55            1.13            1.06          1.05          1.05  
4            1.84            1.71            1.56          0.00          1.05  
...           ...             ...             ...           ...           ...  
1413          NaN             NaN             NaN           NaN           NaN  
1414         0.90            1.15            1.28          3.15          1.05  
1415         2.11            1.08            2.61          2.10          2.10  
1416         0.27            1.04            1.36          3.10          0.00  
1417          NaN             NaN             NaN           NaN           NaN  

[1418 rows x 20 columns]

In [63]:
# We see from the dataframe there are still 2 rows that have NaN values so we can remove them from the dataframe
pl_spi = pl_spi.drop([1413, 1417])

In [64]:
# Check NaNs again
pl_spi.isna().sum()

Date               0
HomeTeam           0
AwayTeam           0
HomeSPI            0
AwaySPI            0
HomeWinProb        0
AwayWinProb        0
TieProb            0
HomeProjScore      0
AwayProjScore      0
HomeImportance    36
AwayImportance    36
score1             0
score2             0
HomeShotXPG        0
AwayShotXPG        0
HomeNonShotXPG     0
AwayNonShotXPG     0
HomeAdjScore       0
AwayAdjScore       0
dtype: int64

We see that there are 36 games for which the importance is unaccounted for.  The importance represents how important
the game is to the respective teams.  Let's say Liverpool are on the verge of a title charge and their opponent has nothing to play for, Liverpool will be given a higher importance value on the match than their opponent.

We can assume that if these values have not been inputted, the match importance is equal for both teams so we can replace the NaN values with 0.

In [65]:
pl_spi.fillna(0, inplace = True)

In [66]:
pl_spi.isna().sum()

Date              0
HomeTeam          0
AwayTeam          0
HomeSPI           0
AwaySPI           0
HomeWinProb       0
AwayWinProb       0
TieProb           0
HomeProjScore     0
AwayProjScore     0
HomeImportance    0
AwayImportance    0
score1            0
score2            0
HomeShotXPG       0
AwayShotXPG       0
HomeNonShotXPG    0
AwayNonShotXPG    0
HomeAdjScore      0
AwayAdjScore      0
dtype: int64

In [67]:
# We will use this to conduct the poisson distribution for the SPI data only.
pl_spi.to_csv(r'pl_spi.csv')

----

### LEGEND

Now that we have our final data table, a legend is provided to help the user understand what each feature represents

- **Date**: Match Date
- **Season**: Premier League Season
- **HomeTeam**: Name of the Home Team
- **AwayTeam**: Name of the Away Team
- **FTHG**: Number of goals scored by the Home Team at Full Time
- **FTAG**: Number of goals scored by the Away Team at Full Time
- **GoalDifference**: Calculated as FTHG - FTAG
- **Referee**: Name of the referee
- **RefBiasScore**: Each referee is assigned a bias score towards the home team
- **HS**: Home Team Shots
- **AS**: Away Team Shots
- **HST**: Home Team Shots on Target
- **AST**: Away Team Shots on Target
- **HF**: Home Team Fouls
- **AF**: Away Team Fouls
- **HC**: Home Team Corners
- **AC**: Away Team Corners
- **HY**: Home Team Yellow Cards
- **AY**: Away Team Yellow Cards
- **HR**: Home Team Red Cards
- **AR**: Away Team Red Cards
- **HomeSPI**: Soccer Power Index for the Home Team
- **AwaySPI**: Soccer Power Index for the Away Team
- **HomeWinProb**: Calculated probability of a Home Win
- **AwayWinProb**: Calculated probability of an Away Win
- **TieProb**: Calculated probability of a draw between the Home and Away Team
- **HomeProjScore**: Projected number of goals scored by the Home Team
- **AwayProjScore**: Projected number of goals scored by the Away Team
- **HomeImportance**: Importance of a win for the Home Team
- **AwayImportance**: Importance of a win for the Away Team
- **HomeShotXPG**: Expected goals based on how many goals the Home Team should have scored, given the shots they took.  Probability is assigned based on distance, angle from the goal, what part of the body the shot was taken with and who took the shot.  The individual shot probabilities are added together to produce this metric which may be larger or smaller than the actual goals scored in the match.
- **AwayShotXPG**: Expected goals based on how many goals the Away Team should have scored, given the shots they took.
- **HomeNonShotXPG**: Non-shot expected goals based on how many goals the home team should have scored dependent on non-shooting actions around the opposition: passes, interceptions, take-ons, tackles.  For example, a completed pass received at the center of the six-yard box leads to a 14% chance of a goal.  The actions added up through the entire match to arrive at this metric and just as the Shot XPG, each action is adjusted based on on the success rate of the player(s).
- **AwayNonShotXPG**: Non-shot expected goals based on how many goals the away team should have scored dependent on non-shooting actions around the opposition: passes, interceptions, take-ons, tackles
- **HomeAdjScore**: Adjusted goals is the value of goals scored for the Home team when a team has more players on the field or late goals if the home team is leading.  Once the goals are downweighted, the value of all other goals are increased to make the total adjusted goals add up as close as possible to the total number of goals scored by the Home Team.
- **AwayAdjScore**: Adjusted goals for the away team.  Calculated similarly to the Home Adjusted Goals Scored.
- **B365H, B365D, B365A**: Bet365 Odds for Home Win, Draw, Away Win respectively
- **FTR**: Full Time Result (represented by the Home Team)